### PR TITLE
add task3/task4

### PR DESCRIPTION
**Pull request by ShulgaVadim**

### CODE REVIEW

File: store/src/main/java/DOMxmlReader.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class DOMxmlReader {
+    public Map<String, String> XMLreader(String path) {
```
Comment (artemmoroz) [inline store/src/main/java/DOMxmlReader.java]: func name (camel case)

File: store/src/main/java/Product.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import lombok.*;
+
+import java.util.Objects;
+
+@Getter
+@Setter
+@AllArgsConstructor
+@Builder
```
Comment (artemmoroz) [inline store/src/main/java/Product.java]: what for builder here?

File: store/src/main/java/ProductPriceComparator.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import java.util.Comparator;
+
+public class ProductPriceComparator implements Comparator<Product> {
+
+
+    @Override
+    public int compare(Product o1, Product o2) {
+        if (o1.getPrice() > o2.getPrice())
```
Comment (artemmoroz) [inline store/src/main/java/ProductPriceComparator.java]: you can use Double.compare instead

File: store/src/main/java/ProductRatingComparator.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import java.util.Comparator;
+
+public class ProductRatingComparator implements Comparator<Product> {
+
+    @Override
+    public int compare(Product o1, Product o2) {
+        if (o1.getRate() > o2.getRate())
```
Comment (artemmoroz) [inline store/src/main/java/ProductRatingComparator.java]: the same as above

File: store/src/main/java/RandomStorePopulator.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import com.github.javafaker.Faker;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+
+public class RandomStorePopulator {
+
+    public ArrayList<Product> populateTheStore() {
```
Comment (artemmoroz) [inline store/src/main/java/RandomStorePopulator.java]: functions should return as minimal possible class/interface, so, here u should use List<Product>

File: store/src/main/java/RandomStorePopulator.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+        Faker faker = new Faker();
```
Comment (artemmoroz) [inline store/src/main/java/RandomStorePopulator.java]: u can init faker instance only once for populator instance or even once for all populator instances

File: store/src/main/java/RandomStorePopulator.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        ArrayList<Product> products = new ArrayList<>();
```
Comment (artemmoroz) [inline store/src/main/java/RandomStorePopulator.java]: the same as at line #8

File: store/src/main/java/Store.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.stream.Collectors;
+
+@AllArgsConstructor
+public class Store {
+
+    ArrayList<Product> products;
```
Comment (artemmoroz) [inline store/src/main/java/Store.java]: List

File: store/src/main/java/Store.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    public void printProducts(ArrayList<Product> products) {
```
Comment (artemmoroz) [inline store/src/main/java/Store.java]: what for u place products as input param here?

File: store/src/main/java/Store.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        for (int i = 0; i < products.size(); i++) {
```
Comment (artemmoroz) [inline store/src/main/java/Store.java]: in this case u can use simplified 'for' syntax ( for(Product p: products)

File: store/src/main/java/RandomStorePopulator.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        for (int i = 0; i <= 15; i++) {
+
+            Category myCategory = null;
```
Comment (artemmoroz) [inline store/src/main/java/RandomStorePopulator.java]: so, all this code is not what was required. The task is to read all Category class implementations in some package, and create their instances and populate some products to them

File: store/src/main/java/Store.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            System.out.println(products.get(i));
+        }
+    }
+
+    public ArrayList<Product> sortByNameAsc() {
```
Comment (artemmoroz) [inline store/src/main/java/Store.java]: please think about all *sort* methods, how they could be rafcatored

File: store/src/main/java/Store.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+                } else if (value.equals("desc")) {
+                    sortedListOfProducts = sortByRatingDesc();
+                }
+                break;
+        }
+
+        return sortedListOfProducts;
+    }
+
+    public void showTopProductsByPriceDesc(int n) {
```
Comment (artemmoroz) [inline store/src/main/java/Store.java]: please read aboud SOLID principle, and split this method in several, and method that get smthing from user (Scanner) should not be in store class

File: store/src/main/java/Store.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        Scanner scanner = new Scanner(System.in);
+        System.out.println("\n Enter 'top' if you want to see 5 products sorted via price desc\n" +
+                " or 'quit' to exit app");
+        String command = scanner.nextLine();
+        if (command.equals("top")) {
+            ArrayList<Product> top5Products = sortByPriceDesc();
+            System.out.println("\n *********sorting top " + n + " products*********\n");
+            for (int i = 0; i < n; i++) {
```
Comment (artemmoroz) [inline store/src/main/java/Store.java]: take a look at java stream api and select proper method from it.

File: store/src/main/java/StoreApp.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+public class StoreApp {
+
+    public static void main(String[] args) {
+
+        RandomStorePopulator populator = new RandomStorePopulator();
+        ArrayList<Product> products = populator.populateTheStore();
+        Store evroOpt = new Store(products);
+        evroOpt.printProducts(products); //print all products from myStore
+
+        String path = System.getProperty("user.dir") + "/store/src/main/resources/sort.xml";
```
Comment (artemmoroz) [inline store/src/main/java/StoreApp.java]: you path are incorrect, cause / is only windows path seporator

File: store/src/main/java/StoreApp.java [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+        DOMxmlReader reader = new DOMxmlReader();
+        Map<String, String> sortCondition = reader.XMLreader(path);
+
+        ArrayList<Product> sortedProducts = evroOpt.sortProduct(sortCondition, "price");
```
Comment (artemmoroz) [inline store/src/main/java/StoreApp.java]: according to the task, sorting should be done according to the whole config file, not just one item. So, create a solution that supports multiple comporators.

**Review by artemmoroz — CHANGES_REQUESTED** [review]
Also, take a look at the project structure, some of the classes should be in domain module, some at console app and so on. parent module is redundant.

File: store/src/main/java/DOMxmlReader.java (hunk 1 above)
Comment (ShulgaVadim) [reply store/src/main/java/DOMxmlReader.java]: changed.

File: store/src/main/java/ProductPriceComparator.java (hunk 3 above)
Comment (ShulgaVadim) [reply store/src/main/java/ProductPriceComparator.java]: changed.

File: store/src/main/java/ProductRatingComparator.java (hunk 4 above)
Comment (ShulgaVadim) [reply store/src/main/java/ProductRatingComparator.java]: changed.

File: store/src/main/java/RandomStorePopulator.java (hunk 5 above)
Comment (ShulgaVadim) [reply store/src/main/java/RandomStorePopulator.java]: changed.

File: store/src/main/java/RandomStorePopulator.java (hunk 7 above)
Comment (ShulgaVadim) [reply store/src/main/java/RandomStorePopulator.java]: changed

File: store/src/main/java/Store.java (hunk 9 above)
Comment (ShulgaVadim) [reply store/src/main/java/Store.java:13]: changed

File: store/src/main/java/Store.java (hunk 8 above)
Comment (ShulgaVadim) [reply store/src/main/java/Store.java]: changed

File: store/src/main/java/Store.java (hunk 10 above)
Comment (ShulgaVadim) [reply store/src/main/java/Store.java]: changed

File: store/src/main/java/Store.java (hunk 9 above)
Comment (ShulgaVadim) [reply store/src/main/java/Store.java]: changed.

File: store/src/main/java/RandomStorePopulator.java (hunk 6 above)
Comment (ShulgaVadim) [reply store/src/main/java/RandomStorePopulator.java]: changed.

File: consoleApp/pom.xml [hunk 17]
```diff
@@ -3,6 +3,14 @@
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
+    <dependencies>
+        <dependency>
+            <groupId>groupId</groupId>
```
Comment (artemmoroz) [inline consoleApp/pom.xml]: invalid name, read about the artifact naming

File: consoleApp/src/main/java/Command.java [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.stream.Collectors;
+
+public class Command {
+
+    public void quit() {
+        System.exit(0);
+    }
+
+    public void showTop(Store store) {
+        List<Product> products = store.products;
```
Comment (artemmoroz) [inline consoleApp/src/main/java/Command.java]: you should use getters and setters, to follow incapsulation principle

File: consoleApp/src/main/java/Command.java [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+        List<Product> topProducts;
+        System.out.println("\n *********sorting top 5 products*********\n");
+        Comparator<Product> productPriceComparatorDesc = Comparator.comparing(Product::getPrice, Comparator.reverseOrder());
+        topProducts = products.stream().sorted(productPriceComparatorDesc).limit(5).collect(Collectors.toList());
```
Comment (artemmoroz) [inline consoleApp/src/main/java/Command.java]: it is  not an error, but it is better to parameterize such functions, and set the number of products as income param

File: consoleApp/src/main/java/Command.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+        topProducts.forEach(System.out::println);
```
Comment (artemmoroz) [inline consoleApp/src/main/java/Command.java]: you can print products in the above stream

File: consoleApp/src/main/java/Command.java [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    public void getCommandForStore (Store store) {
+        Scanner scanner = new Scanner(System.in);
+        System.out.println("\n Enter 'top' if you want to see 5 products sorted via price desc\n for "
+                + " or 'quit' to exit app");
+        String command = scanner.nextLine();
+        if (command.equals("top")) {
```
Comment (artemmoroz) [inline consoleApp/src/main/java/Command.java]: i think there should be sort command also, cause you need to handle : simple print products, sorted, top and quit

File: consoleApp/src/main/java/StoreApp.java [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+
+    public static void main(String[] args) {
+        RandomStorePopulator populator = new RandomStorePopulator();
+        List<Product> products = populator.populateTheStore();
+        Store evroOpt = new Store(products);
+        evroOpt.printProducts(); //print all products from myStore
+
+        URL resource = StoreApp.class.getResource("sort.xml");
+        DomXmlReader reader = new DomXmlReader();
+        Map<String, String> sortCondition = reader.XmlReader(resource.getPath());
```
Comment (artemmoroz) [inline consoleApp/src/main/java/StoreApp.java]: XmlReader - func name, camel case!

File: domain/src/main/java/categories/Fruit.java [hunk 23]
```diff
@@ -0,0 +1,12 @@
+package categories;
+
+
+import product.Category;
+
+public class Fruit extends Category {
+    private static final String FRUIT_CATEGORY = "fruit";
+
+    public Fruit(){
+        setName(FRUIT_CATEGORY);
```
Comment (artemmoroz) [inline domain/src/main/java/categories/Fruit.java]: super constructor is preferable

File: store/src/main/java/RandomStorePopulator.java [hunk 24]
```diff
@@ -1,33 +1,63 @@
...
-import java.lang.reflect.InvocationTargetException;
 import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 public class RandomStorePopulator {
 
-    public ArrayList<Product> populateTheStore() {
+    Reflections reflections = new Reflections("categories");
```
Comment (artemmoroz) [inline store/src/main/java/RandomStorePopulator.java]: wrong package name

File: store/src/main/java/RandomStorePopulator.java [hunk 25]
```diff
@@ -1,33 +1,63 @@
...
+    List<String> namesOfCategory = new ArrayList<>();
+    Set<Class<? extends Category>> subClassesForCategory = reflections.getSubTypesOf(Category.class);
+    Faker faker = new Faker();
 
-        Faker faker = new Faker();
-        ArrayList<Product> products = new ArrayList<>();
-        for (int i = 0; i <= 15; i++) {
+    public List<String> getNamesOfCategory() {
+        for (int j = 0; j < subClassesForCategory.size(); j++) {
+            namesOfCategory.add(subClassesForCategory.stream().collect(Collectors.toList())
```
Comment (artemmoroz) [inline store/src/main/java/RandomStorePopulator.java]: very strange method, and a lot of cpu overhead, you read category list multiple times.

File: store/src/main/java/RandomStorePopulator.java [hunk 26]
```diff
@@ -1,33 +1,63 @@
...
+    }
 
-            Category myCategory = null;
+    public List<Product> populateTheStore() {
+        List<Product> products = new ArrayList<>();
+        String productName;
+        String className;
+
+        for (int i = 0; i < 20; i++) {
+            int randomInt = (int) (Math.random() * subClassesForCategory.size());
```
Comment (artemmoroz) [inline store/src/main/java/RandomStorePopulator.java]: please, think one more time about this method impl. it is very complex and overengineered.

File: store/src/main/java/Store.java [hunk 27]
```diff
@@ -1,112 +1,59 @@
...
-                }
-                break;
-            case "rate":
-                if (value.equals("asc")) {
-                    sortedListOfProducts = sortByRatingAsc();
-                } else if (value.equals("desc")) {
-                    sortedListOfProducts = sortByRatingDesc();
-                }
-                break;
+    public List<Product>multipleSort(Map<String, String> map){
```
Comment (artemmoroz) [inline store/src/main/java/Store.java]: a lot of duplicated code, and multiple sort still not supported.

File: store/src/main/java/utility/DomXmlReader.java [hunk 28]
```diff
@@ -11,8 +13,8 @@
 import java.util.HashMap;
 import java.util.Map;
 
-public class DOMxmlReader {
-    public Map<String, String> XMLreader(String path) {
+public class DomXmlReader {
+    public Map<String, String> XmlReader(String path) {
```
Comment (artemmoroz) [inline store/src/main/java/utility/DomXmlReader.java]: https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html  - CAMEL CASE

**Review by artemmoroz — CHANGES_REQUESTED** [review]

**Comment by artemmoroz** [comment]
everything is fine

**Pull request merged**